src/libraries/Common/tests/System/Security/Cryptography/MLKemTests.Encapsulation.cs

```diff
@@ -0,0 +1,253 @@
+
+        [ConditionalFact(typeof(MLKem), nameof(MLKem.IsSupported))]
+        public static void DecapsulateVectors_ModifiedCiphertext()
+        {
+            foreach (MLKemTestDecapsulationVector vector in MLKemDecapsulationTestVectors)
+            {
+                byte[] decapsulationKeyBytes = vector.DecapsulationKey.HexToByteArray();
+                using MLKem kem = MLKem.ImportDecapsulationKey(vector.Algorithm, decapsulationKeyBytes);
+
+                byte[] sharedSecretBuffer = new byte[MLKem.SharedSecretSizeInBytes];
+                byte[] expectedSharedSecret = vector.SharedSecret.HexToByteArray();
+                byte[] ciphertext = vector.Ciphertext.HexToByteArray();
+
+                Tamper(ciphertext);
+                kem.Decapsulate(ciphertext, sharedSecretBuffer);
+
+                AssertExtensions.SequenceNotEqual(expectedSharedSecret, sharedSecretBuffer);
+            }
+        }
+
+        [ConditionalTheory(typeof(MLKem), nameof(MLKem.IsSupported))]
+        [MemberData(nameof(MLKemAlgorithms))]
+        public static void Encapsulate_NonDeterministic(MLKemAlgorithm algorithm)
+        {
+            using MLKem kem = MLKem.GenerateKey(algorithm);
+
+            byte[] ciphertext1 = new byte[algorithm.CiphertextSizeInBytes];
+            byte[] sharedSecret1 = new byte[MLKem.SharedSecretSizeInBytes];
+            byte[] ciphertext2 = new byte[algorithm.CiphertextSizeInBytes];
+            byte[] sharedSecret2 = new byte[MLKem.SharedSecretSizeInBytes];
+            kem.Encapsulate(ciphertext1, sharedSecret1);
+            kem.Encapsulate(ciphertext2, sharedSecret2);
+            AssertExtensions.SequenceNotEqual(ciphertext1, ciphertext2);
+            AssertExtensions.SequenceNotEqual(sharedSecret1, sharedSecret2);
+        }
+
+        [ConditionalTheory(typeof(MLKem), nameof(MLKem.IsSupported))]
+        [InlineData(false)]
+        [InlineData(true)]
+        public static void Encapsulate_Overlaps(bool partial)
+        {
+            using MLKem kem = MLKem.GenerateKey(MLKemAlgorithm.MLKem512);
+            byte[] buffer = new byte[MLKemAlgorithm.MLKem512.CiphertextSizeInBytes];
+
+            Assert.Throws<CryptographicException>(() =>
+            {
+                Span<byte> sharedSecret = buffer.AsSpan(partial ? 1 : 0, MLKem.SharedSecretSizeInBytes);
+                Span<byte> ciphertext = buffer.AsSpan(0, MLKemAlgorithm.MLKem512.CiphertextSizeInBytes);
+                kem.Encapsulate(ciphertext, sharedSecret);
+            });
+        }
+
+        private static void Tamper(Span<byte> buffer)
+        {
+            buffer[buffer.Length - 1] ^= 0xFF;
+        }
+
+        public record MLKemTestDecapsulationVector(MLKemAlgorithm Algorithm, string EncapsulationKey, string DecapsulationKey, string Ciphertext, string SharedSecret);
+
+        public static IEnumerable<MLKemTestDecapsulationVector> MLKemDecapsulationTestVectors
+        {
+            get
+            {
+                yield return new
```

Comment:
I feel like in other algorithms when we are excerpting the CAVP/ACVP tests we cite them per vector-set.  More or less just indicating a difference between "this came from a standard test set" and "I did this using the openssl commandline" (or whatever).

Reply:
Ah, I did that for the key gen, but not for encapsulation.